pkg/controllers/management/auth/globalroles/globalrole_handler.go

```diff
@@ -196,6 +217,110 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
+func (gr *globalRoleLifecycle) reconcileNamespacedRoles(globalRole *v3.GlobalRole) error {
+	var returnError error
+
+	// For collecting all the roles that should exist for the GlobalRole
+	roleUIDs := []types.UID{}
+
+	for ns, rules := range globalRole.NamespacedRules {
+		_, err := gr.nsLister.Get("", ns)
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get namespace %s", ns))
+			continue
+		}
+		// Check if the role exists
+		roleName := globalRole.Name + "-" + ns
+		role, err := gr.rLister.Get(ns, roleName)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				returnError = multierror.Append(returnError, err)
+				continue
+			}
+			newRole := &v1.Role{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      roleName,
+					Namespace: ns,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: globalRole.APIVersion,
+							Kind:       globalRole.Kind,
+							Name:       globalRole.Name,
+							UID:        globalRole.UID,
+						},
+					},
+					Labels: map[string]string{
+						grOwnerLabel: wrangler.SafeConcatName(globalRole.Name),
+					},
+				},
+				Rules: rules,
+			}
+			createdRole, err := gr.rClient.Create(newRole)
+			if err == nil {
+				roleUIDs = append(roleUIDs, createdRole.UID)
+				continue
+			} else if !apierrors.IsAlreadyExists(err) {
+				returnError = multierror.Append(returnError, err)
+				continue
+			} else {
+				// In the case that the role already exists, we fall through to check that the rules are correct
+				role = newRole
+			}
+		}
+		if role != nil {
+			roleUIDs = append(roleUIDs, role.GetUID())
+
+			// Check that the rules for the existing role are correct and that it has the right Owner Label
+			if !reflect.DeepEqual(role.Rules, rules) || role.Labels[grOwnerLabel] != globalRole.Name {
+				newRole := role.DeepCopy()
+				newRole.Rules = rules
+				if newRole.Labels != nil {
+					newRole.Labels[grOwnerLabel] = globalRole.Name
+				} else {
+					newRole.Labels = map[string]string{grOwnerLabel: globalRole.Name}
+				}
+				_, err := gr.rClient.Update(newRole)
+				if err != nil {
+					returnError = multierror.Append(returnError, err)
+					continue
+				}
+			}
+		}
+	}
+
+	// get all the roles claiming to be owned by this GR and remove any that shouldn't exist
+	r, err := labels.NewRequirement(grOwnerLabel, selection.Equals, []string{globalRole.Name})
+	if err != nil {
+		returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't create label: %s", grOwnerLabel))
+	} else {
+		roles, err := gr.rLister.List("", labels.NewSelector().Add(*r))
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't list roles with label %s : %s", grOwnerLabel, globalRole.Name))
+		} else {
+			err = gr.purgeInvalidNamespacedRoles(roles, roleUIDs)
+			if err != nil {
+				returnError = multierror.Append(returnError, err)
+			}
+		}
+	}
```

Comment:
Maybe simply flow by removing nested ifs:

```suggestion
	if err != nil {
		return multierror.Append(returnError, errors.Wrapf(err, "couldn't create label: %s", grOwnerLabel))
	}
	
	roles, err := gr.rLister.List("", labels.NewSelector().Add(*r))
	if err != nil {
		return multierror.Append(returnError, errors.Wrapf(err, "couldn't list roles with label %s : %s", grOwnerLabel, globalRole.Name))
	}
	
	err = gr.purgeInvalidNamespacedRoles(roles, roleUIDs)
	if err != nil {
		returnError = multierror.Append(returnError, err)
	}
```